kalite/coachreports/api_views.py

```diff
@@ -1,382 +1,87 @@
-    ex_logs = dict(zip([w.id for w in who], [[] for i in range(len(who))]))
-    if UserLog.is_enabled():
-        activity_logs = dict(zip([w.id for w in who], [[] for i in range(len(who))]))
-
-    # Set up queries (but don't run them), so we have really easy aliases.
-    #   Only do them if they haven't been done yet (tell this by passing in a value to the lambda function)
-    # Topics: topics.
-    # Exercises: names (ids for ExerciseLog objects)
-    # Videos: video_id (ids for VideoLog objects)
-
-    # This lambda partial creates a function to return all items with a particular path from the NODE_CACHE.
-    search_fun_single_path = partial(lambda t, p: t["path"].startswith(p), p=tuple(where))
-    # This lambda partial creates a function to return all items with paths matching a list of paths from NODE_CACHE.
-    search_fun_multi_path = partial(lambda ts, p: any([t["path"].startswith(p) for t in ts]),  p=tuple(where))
-    # Functions that use the functions defined above to return topics, exercises, and videos based on paths.
-    query_exercises = partial(lambda e, sf: e if e is not None else [ex["id"] for ex in filter(sf, get_exercise_cache(language=language).values())], sf=search_fun_single_path)
-    query_videos = partial(lambda v, sf: v if v is not None else [vid["id"] for vid in filter(sf, get_node_cache('Content', language=language).values())], sf=search_fun_single_path)
-
-    # No users, don't bother.
-    if len(who) > 0:
-
-        # Query out all exercises, videos, exercise logs, and video logs before looping to limit requests.
-        # This means we could pull data for n-dimensional coach report displays with the same number of requests!
-        # Note: User activity is polled inside the loop, to prevent possible slowdown for exercise and video reports.
-        exercises = query_exercises(exercises)
+        # Do this to ensure that we never return more than one facility's worth of anything.
+        learner_filter = Q(facility__pk__in=facility_ids)
 
-        videos = query_videos(videos)
+    topic_ids = request.GET.getlist("topic_id", [])
 
-        if exercises:
-            ex_logs = query_logs(data.keys(), exercises, "exercise", ex_logs)
+    learners = FacilityUser.objects.filter(learner_filter).order_by("last_name")
 
-        if videos:
-            vid_logs = query_logs(data.keys(), videos, "video", vid_logs)
+    pages = int(ceil(len(learners)/float(limit)))
 
-        for data_type in (data_types if not hasattr(data_types, "lower") else [data_types]):  # convert list from string, if necessary
-            if data_type in data[data.keys()[0]]:  # if the first user has it, then all do; no need to calc again.
-                continue
+    if page*limit < len(learners):
 
-            #
-            # These are summary stats: you only get one per user
-            #
-            if data_type == "pct_mastery":
+        learners = learners[(page - 1)*limit: page*limit]
```

Comment:
If you over count the end of an index range (i.e. if page*limit is bigger than the length of the list) it just returns everything to the end of the list. For the sake of simplicity, I left it like this so that we could use the same logic regardless of which page was being requested.